File: src/api/library.ts
import { commands } from './commands'
import { executeCommand, listResult } from './client'
import type {
  Axiom,
  AxiomDraft,
  Character,
  CharacterDraft,
  Entry,
  EntryDraft,
  EntityRef,
  EventDraft,
  EventParticipantDraft,
  EventRecord,
  LibraryProjectSnapshot,
  Relation,
  RelationDraft,
} from '@/types/library'

export function loadProjectSnapshot(projectId: string): Promise<LibraryProjectSnapshot> {
  return executeCommand(commands.library.projectSnapshot, { projectId })
}

export function createEntry(draft: EntryDraft): Promise<Entry> {
  return executeCommand(commands.library.createEntry, { draft })
}

export function updateEntry(id: string, draft: EntryDraft): Promise<Entry> {
  return executeCommand(commands.library.updateEntry, { id, draft })
}

export function deleteEntry(id: string): Promise<void> {
  return executeCommand(commands.library.deleteEntry, { id })
}

export function createCharacter(draft: CharacterDraft): Promise<Character> {
  return executeCommand(commands.library.createCharacter, { draft })
}

export function updateCharacter(id: string, draft: CharacterDraft): Promise<Character> {
  return executeCommand(commands.library.updateCharacter, { id, draft })
}

export function deleteCharacter(id: string): Promise<void> {
  return executeCommand(commands.library.deleteCharacter, { id })
}

export function createEvent(
  draft: EventDraft,
  participants: EventParticipantDraft[] = [],
): Promise<EventRecord> {
  return executeCommand(commands.library.createEvent, { draft, participants })
}

export function updateEvent(
  id: string,
  draft: EventDraft,
  participants: EventParticipantDraft[] = [],
): Promise<EventRecord> {
  return executeCommand(commands.library.updateEvent, { id, draft, participants })
}

export function deleteEvent(id: string): Promise<void> {
  return executeCommand(commands.library.deleteEvent, { id })
}

export function createAxiom(draft: AxiomDraft): Promise<Axiom> {
  return executeCommand(commands.library.createAxiom, { draft })
}

export function updateAxiom(id: string, draft: AxiomDraft): Promise<Axiom> {
  return executeCommand(commands.library.updateAxiom, { id, draft })
}

export function deleteAxiom(id: string): Promise<void> {
  return executeCommand(commands.library.deleteAxiom, { id })
}

export function listBacklinks(target: EntityRef): Promise<Relation[]> {
  return executeCommand(commands.library.listBacklinks, { target }).then(listResult)
}

export function createRelation(draft: RelationDraft): Promise<Relation> {
  return executeCommand(commands.library.createRelation, { draft })
}

export function updateRelation(id: string, draft: RelationDraft): Promise<Relation> {
  return executeCommand(commands.library.updateRelation, { id, draft })
}

export function deleteRelation(id: string): Promise<void> {
  return executeCommand(commands.library.deleteRelation, { id })
}

File: src-tauri/src/commands/library.rs
use tauri::State;

use crate::{
    db::app_state::AppState,
    domain::{
        axiom::{Axiom, AxiomDraft, AxiomRepository},
        character::{Character, CharacterDraft, CharacterRepository},
        entry::{Entry, EntryDraft, EntryRepository},
        event::{Event, EventDraft, EventParticipantDraft, EventRepository},
        relation::{EntityRef, Relation, RelationDraft, RelationRepository},
    },
    services::library::{load_project_snapshot, LibraryProjectSnapshot},
};

#[tauri::command]
pub fn library_project_snapshot(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<LibraryProjectSnapshot, String> {
    state.with_database(|connection| load_project_snapshot(connection, &project_id))
}

#[tauri::command]
pub fn library_create_entry(
    state: State<'_, AppState>,
    draft: EntryDraft,
) -> Result<Entry, String> {
    state.with_database(|connection| EntryRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn library_update_entry(
    state: State<'_, AppState>,
    id: String,
    draft: EntryDraft,
) -> Result<Entry, String> {
    state.with_database(|connection| EntryRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn library_delete_entry(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| EntryRepository::new(connection).soft_delete(&id))
}

#[tauri::command]
pub fn library_create_character(
    state: State<'_, AppState>,
    draft: CharacterDraft,
) -> Result<Character, String> {
    state.with_database(|connection| CharacterRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn library_update_character(
    state: State<'_, AppState>,
    id: String,
    draft: CharacterDraft,
) -> Result<Character, String> {
    state.with_database(|connection| CharacterRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn library_delete_character(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| CharacterRepository::new(connection).soft_delete(&id))
}

#[tauri::command]
pub fn library_create_event(
    state: State<'_, AppState>,
    draft: EventDraft,
    participants: Vec<EventParticipantDraft>,
) -> Result<Event, String> {
    state.with_database(|connection| EventRepository::new(connection).create(draft, participants))
}

#[tauri::command]
pub fn library_update_event(
    state: State<'_, AppState>,
    id: String,
    draft: EventDraft,
    participants: Vec<EventParticipantDraft>,
) -> Result<Event, String> {
    state.with_database(|connection| {
        EventRepository::new(connection).update(&id, draft, participants)
    })
}

#[tauri::command]
pub fn library_delete_event(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| EventRepository::new(connection).soft_delete(&id))
}

#[tauri::command]
pub fn library_create_axiom(
    state: State<'_, AppState>,
    draft: AxiomDraft,
) -> Result<Axiom, String> {
    state.with_database(|connection| AxiomRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn library_update_axiom(
    state: State<'_, AppState>,
    id: String,
    draft: AxiomDraft,
) -> Result<Axiom, String> {
    state.with_database(|connection| AxiomRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn library_delete_axiom(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| AxiomRepository::new(connection).soft_delete(&id))
}

#[tauri::command]
pub fn library_list_backlinks(
    state: State<'_, AppState>,
    target: EntityRef,
) -> Result<Vec<Relation>, String> {
    state.with_database(|connection| RelationRepository::new(connection).list_backlinks(&target))
}

#[tauri::command]
pub fn library_create_relation(
    state: State<'_, AppState>,
    draft: RelationDraft,
) -> Result<Relation, String> {
    state.with_database(|connection| RelationRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn library_update_relation(
    state: State<'_, AppState>,
    id: String,
    draft: RelationDraft,
) -> Result<Relation, String> {
    state.with_database(|connection| RelationRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn library_delete_relation(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| RelationRepository::new(connection).soft_delete(&id))
}
